Implement Steam Authentication related APIs

jahrain opened the pull request:
I am using greenworks for porting an online game over to Steam and implemented some of the authentication related APIs required for it. There may be some unrelated commits in this PR, so feel free to rebase or merge these changes as needed.

hokein commented on src/greenworks_api.cc:
```diff
@@ -587,6 +587,57 @@ NAN_METHOD(ExtractArchive) {
+  NanReturnUndefined();
+}
+
+
```
Remove the extra blank line.

hokein commented on src/greenworks_api.cc:
```diff
@@ -587,6 +587,57 @@ NAN_METHOD(ExtractArchive) {
+  if (args.Length() < 2 || !args[1]->IsFunction()) {
+    THROW_BAD_ARGS("Bad arguments");
+  }
+  v8::String::Utf8Value _user_data(args[0]->ToString());
```
The `_user_data` name isn't aligned with code style, Greenworks uses [Google C++ code style](https://google-styleguide.googlecode.com/svn/trunk/cppguide.html)

hokein commented on src/greenworks_api.cc:
```diff
@@ -587,6 +587,57 @@ NAN_METHOD(ExtractArchive) {
+  std::string user_data = *_user_data;
+  NanCallback* success_callback = new NanCallback(args[1].As<v8::Function>());
+  NanCallback* error_callback = NULL;
+  if (args.Length() > 1 && args[1]->IsFunction())
```
Shouldn't check `args[2]` here for `error_callback` ? since the `args[1]` is for `success_callback`.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+void GetAuthSessionTicketWorker::OnGetAuthSessionCompleted(GetAuthSessionTicketResponse_t *inCallback) {
+  if (inCallback->m_eResult == k_EResultOK) {
+    std::ostringstream hex_ticket;
+    for (unsigned int i = 0; i < ticket_buf_size_; i++) {
```
Use `++i` instead.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+  if (inCallback->m_eResult == k_EResultOK) {
+    std::ostringstream hex_ticket;
+    for (unsigned int i = 0; i < ticket_buf_size_; i++) {
+        hex_ticket << std::setfill('0') << std::setw(2) << std::uppercase << std::hex << int(ticket_buf_[i]);
```
Could you add a comment to illustrate the format code of `hex_ticket`? It's not easy to understand to format at first glance.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+
+
+
+
```
Remove the extra blank lines.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+}
+
+void RequestEncryptedAppTicketWorker::Execute() {
+  SteamAPICall_t steam_api_call = SteamUser()->RequestEncryptedAppTicket((void*)user_data_.c_str(), user_data_.length());
```
Use `static_cast` for type cast rather than C-style.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+  if (!io_failure && inCallback->m_eResult == k_EResultOK) {
+    SteamUser()->GetEncryptedAppTicket(ticket_buf_, sizeof(ticket_buf_), &ticket_buf_size_);
+    std::ostringstream hex_ticket;
+    for (unsigned int i = 0; i < ticket_buf_size_; i++) {
```
Use `++i` instead.

hokein commented on src/greenworks_async_workers.cc:
```diff
@@ -261,4 +264,78 @@ void ExtractArchiveWorker::Execute() {
+    SteamUser()->GetEncryptedAppTicket(ticket_buf_, sizeof(ticket_buf_), &ticket_buf_size_);
+    std::ostringstream hex_ticket;
+    for (unsigned int i = 0; i < ticket_buf_size_; i++) {
+      hex_ticket << std::setfill('0') << std::setw(2) << std::uppercase << std::hex << int(ticket_buf_[i]);
```
The same as above. I think you can put this formating code into a utility function that defined in `anonymous namespace` of this file, so that we don't maintain two copies of that.

hokein commented on src/greenworks_async_workers.h:
```diff
@@ -162,6 +162,41 @@ class ExtractArchiveWorker : public SteamAsyncWorker {
+  unsigned char ticket_buf_[2048];
+};
+
+
```
Remove the blank line.

hokein commented on src/greenworks_async_workers.h:
```diff
@@ -162,6 +162,41 @@ class ExtractArchiveWorker : public SteamAsyncWorker {
+  CCallResult< RequestEncryptedAppTicketWorker, EncryptedAppTicketResponse_t > call_result_;
+};
+    
+    
```
Also remove the extra blank line.

hokein commented:
@jahrain 
Thanks for your contributions on Greenworks! You PR looks great roughly, and I have left some comments for more improvements. 

BTW, could you please also add the the APIs definitions like the ones in [`wiki`](https://github.com/greenheartgames/greenworks/wiki/Cloud-APIs), so that make the this PR more understable?

I have no practice experience on Steam Authentication APIs, is it possible to provide a way to test it just like the ` greenworks.saveTextToFile` API in [test sample](https://github.com/greenheartgames/greenworks/blob/master/test/test.js)?

jahrain commented:
@hokein 
I added the Authentication API's to the Wiki https://github.com/greenheartgames/greenworks/wiki/Steam-Authentication-APIs. The official documentation of these APIs are here: https://partner.steamgames.com/documentation/auth. This PR only implements the client-side authentication functions that create tickets which can be used by game servers to authenticate users. 

Also thanks for your detailed code review. In addition, I ran cpplint on the code and fixed any style errors reported as well as made the fixes you suggested. Though I don't have too much time to further improve upon this PR for now, it would be very helpful for other contributors to include a CONTRIBUTING.md file outlining the code formatting rules for PRs.

hokein commented:
@jahrain 
Great! Thanks for your hard work here. 

The PR looks good to me now, I merge it and will make a release in the future.

patrickklug commented:
@jahrain thanks for making this contribution!
